planner/core/logical_plans.go

```diff
@@ -258,18 +275,59 @@ func (p *LogicalJoin) extractFDForOuterJoin() *fd.FDSet {
 
 	filterFD := &fd.FDSet{HashCodeToUniqueID: make(map[string]int)}
 
-	constUniqueIDs := extractConstantCols(eqCondSlice, p.SCtx(), filterFD)
+	constUniqueIDs := extractConstantCols(allConds, p.SCtx(), filterFD)
 
-	equivUniqueIDs := extractEquivalenceCols(eqCondSlice, p.SCtx(), filterFD)
+	equivUniqueIDs := extractEquivalenceCols(allConds, p.SCtx(), filterFD)
 
 	filterFD.AddConstants(constUniqueIDs)
+	equivLeftUniqueIDs := fd.NewFastIntSet()
+	equivAcrossNum := 0
 	for _, equiv := range equivUniqueIDs {
 		filterFD.AddEquivalence(equiv[0], equiv[1])
+		if equiv[0].SubsetOf(outerCols) && equiv[1].SubsetOf(innerCols) {
+			equivLeftUniqueIDs.UnionWith(equiv[0])
+			equivAcrossNum++
+			continue
+		}
+		if equiv[0].SubsetOf(innerCols) && equiv[1].SubsetOf(outerCols) {
+			equivLeftUniqueIDs.UnionWith(equiv[1])
+			equivAcrossNum++
+		}
 	}
 	filterFD.MakeNotNull(notNullColsFromFilters)
 
+	// pre-perceive the filters for the convenience judgement of 3.3.1.
+	var opt fd.ArgOpts
+	if equivAcrossNum > 0 {
+		// find the equivalence FD across left and right cols.
+		var leftConditionCols []*expression.Column
+		if len(p.LeftConditions) != 0 {
```

Comment:
This won't work for right join. We can use innerCondition and outerCondition instead to make it clear.